### PR TITLE
[SPARK-41518][SQL] Assign a name to the error class `_LEGACY_ERROR_TEMP_2422`

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to assign new name `MISSING_GROUP_BY` to the legacy error class `_LEGACY_ERROR_TEMP_2422`, improve its error message format, and regenerate the SQL golden files.

### Why are the changes needed?
To improve user experience with Spark SQL, and unify representation of error messages.

### Does this PR introduce _any_ user-facing change?
Yes, it changes an user-facing error message.

### How was this patch tested?
By running the affected test suites:
```
$ PYSPARK_PYTHON=python3 build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite"
```

### CODE REVIEW

File: core/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -496,6 +496,11 @@
     ],
     "sqlState" : "42000"
   },
+  "GROUPING_EXPRESSIONS_NOT_FOUND" : {
+    "message" : [
+      "Not found any GROUP BY expressions, and <sqlExpr> is not an aggregate function. Wrap <aggExprs> by windowing function(s) or wrap <sqlExpr> by `first()` (or `first_value()`) if you don't care which value you get."
+    ]
+  },
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:503]: @srielau Are you ok with the error class and message format?

Comment (srielau) [reply core/src/main/resources/error/error-classes.json:503]: What is more likely? The user forgot the GROUP BY clause, the user forgot to aggregate the expression, or they meant to use window functions? My money is on forgetting the GROUP BY clause....
How about:
GROUPING_EXPRESSION_WITHOUT_GROUP_BY
The query including grouping expression <groupingExpr> and aggregate expressions <aggExprs> does not include a GROUP BY clause. Add "GROUP BY <groupingExpr>", aggregate <groupingExpr>, or turn the <aggExprs> into window functions using OVER clauses.

File: sql/core/src/test/resources/sql-tests/results/group-by.sql.out [hunk 2]
```diff
@@ -355,10 +354,9 @@ struct<>
 -- !query output
 org.apache.spark.sql.AnalysisException
 {
-  "errorClass" : "_LEGACY_ERROR_TEMP_2422",
+  "errorClass" : "AGGREGATION_WITHOUT_GROUP_BY",
   "messageParameters" : {
-    "aggExprs" : "()",
-    "sqlExpr" : "id"
+    "aggExprs" : ""
```
Comment (MaxGekk) [inline sql/core/src/test/resources/sql-tests/results/group-by.sql.out:359]: Here is the case when there are no aggregations at all:
```sql
SELECT id FROM range(10) HAVING id > 0
```
How about just `MISSING_GROUP_BY`?

**Comment by MaxGekk** [comment]
@srielau @itholic @cloud-fan @panbingkun @LuciferYang Please, review this PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 3]
```diff
@@ -398,15 +398,9 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog with QueryErrorsB
                   }
                 }
               case e: Attribute if groupingExprs.isEmpty =>
```
Comment (LuciferYang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:400]: Seems `e` is no longer used, replace with `_` ?

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: replaced

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 4]
```diff
@@ -397,16 +397,10 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog with QueryErrorsB
...
-                }.nonEmpty)
-                e.failAnalysis(
-                  errorClass = "_LEGACY_ERROR_TEMP_2422",
-                  messageParameters = Map(
-                    "sqlExpr" -> e.sql,
-                    "aggExprs" -> aggExprs.map(_.sql).mkString("(", ", ", ")")))
+              case _: Attribute if groupingExprs.isEmpty =>
+                operator.failAnalysis(
+                  errorClass = "MISSING_GROUP_BY",
+                  messageParameters = Map.empty)
```
Comment (itholic) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:403]: nit: can we clarify the type with `Map.empty[String, String]` To explicitly tell which types can come ?

**Review by itholic — COMMENTED** [review]
Looks pretty good otherwise

**Review by itholic — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 4 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:403]: Are there any reasons to do so in OSS?

**Review by LuciferYang — APPROVED** [review]
+1,LGTM

**Comment by MaxGekk** [comment]
Merging to master. Thank you, @itholic @LuciferYang for review.

**Pull request closed without merging**